site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -0,0 +1,200 @@
+Citra has supported Geometry Shaders since the summer of 2016 thanks to the heroic efforts of the many people that have worked on it in the past.
+
+When trying to convert the PICA200 geometry shaders into GLSL, phantom ran into some very tough problems.
+Some games rely on the shader state carrying over from run to run, which isn't something that you can do in GLSL.
+Even more so, some games rely on (i know i'm forgetting something here??????).
+As a tradeoff, Citra now supports hardware geometry shaders only when the game is configured to use one specific mode, and will seamlessly fall back to CPU geometry shaders if the mode is not supported.
+
+### Multiplication Shouldn't Be This Slow
+
+When converting from PICA200 shaders into GLSL, there's a few PICA200 opcodes that should just match up without any issues.
+Addition, subtraction, and multiplication should ... wait. Where did this issue come from?
+
+{{< sidebyside "image" "/images/entry/improvements-to-hardware-renderer/"
+    "accurate_mul_on.png=On: I don't see any issue..." 
+    "accurate_mul_off.png=Off: Oh. Well maybe Eren looks better this way" >}}
+
+It turns out that the PICA200 multiplication opcode has a few edge cases that doesn't impact a large majority of games, and leads to some hilarious results in others.
+On the PICA200, `infinity * 0 = 0` but in OpenGL `infinity * 0 = NaN` and this can't be configured.
+In the generated GLSL shaders, phantom emulates this behavior by making a function call instead of a simple multiplication.
+
+```glsl
+vec4 sanitize_mul(vec4 lhs, vec4 rhs) {
+    vec4 product = lhs * rhs;
+    return mix(product, mix(mix(vec4(0.0), product, isnan(rhs)), product, isnan(lhs)), isnan(product));
+}
+```
+
+Alas, it's a performance penalty to use a function everywhere instead of regular multiplication.
+On weaker GPUs, we noticed the penalty is so severe, we actually made this configurable.
+The whole point of a hardware renderer is to be fast, so eating a penalty when only a small handful of games need this level of accuracy would be regrettable.
+You can turn off this feature in the settings by deselecting "Accurate Shader" and get a noticeable performance boost, but be aware that a few games will break in strange ways!
+
+### Finding Bugs and Working Overtime
+
+We were very excited to launch this feature when phantom declared that it was ready; results from user testing was entirely positive, and the performance improvements were unbelievable, but one thing stood in the way.
+No one had yet tested to see if it worked on AMD GPUs.
+We called for our good friend [JMC47](https://dolphin-emu.org/blog/authors/JMC47/) to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! Oh no!
+
+From JMC47's time in Dolphin, he's made a few friends here and there, and he found someone willing to investigate.
+After a few grueling weeks, this developer was able to narrow down what the problem is, and luckily it's not a bug in the AMD drivers.
```

Comment:
Is it worth naming the dev who helped us out to thank them?

Reply:
i just barely learned who it was today ^^;